### ISSUE TITLE
embedded mak:list correlation using IN (results of outer query)

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#473]
**Reported by @cristianbogdan on 18 Jan 2006 03:28 UTC**
with left joins, subqueries and all goodies in HQL, it should be possible to
implement a mak:list groups by one single left-join query instead of N+M (where
N= the number of mak:lists and M= the number of nullable pointers)

this would also fix bug 56

Migrated-From: http://trac.makumba.org/ticket/923

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 Jan 2006 03:31 UTC**
the classes org.makumba.view ComposedQuery and Grouper will have to be re-done.
it may be possilble that we go through query results element by element instead
of reading the whole result set as Grouper does now (takes more result sets and
compares them). this should lead to higher performance and lower memory profile

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 19 Jan 2006 11:59 UTC**
bug 916 should be fixed before this and then there will be testing material

this bug will also make redundant bug 50 and probably others

besides the classes mentioned in comment 0,
org.makumba.view.jsptaglib.QueryExecution and maybe others seem to need changes
because the idea of "one query per mak:list" is quite deep in the analysis of
org.makumba.view and org.makumba.view.jsptaglib. still many of the concepts from
the analysis will stay on
http://www.makumba.org/hacking/javadoc/org/makumba/view/jsptaglib/package-summary.html#package_description

basically i think we have to replace "new query" with "left join" both for inner
mak:lists and for nullable pointers (nullableValue in the doc above). so we'll
have no more nullableValueQueries for example.

this bug is easy to fix by non-cristi people because it involves only a handful
of classes whose analysis is quite well documented. i guess a good first step
(except for writing some tests) is to start re-writing the analysis document
above with a left join perspective

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Oct 2007 21:08 UTC**
one way to fix this is not to use LEFT JOIN and one single query but still N queries with IN SET like

first query defining labels a, b

second query WHERE a IN (results of first query) AND b in (results of first query)

currently, the condition of the first query is repeated in the second

this will result in longer query text but most probably faster queries since the conditions in the first query may include subqueries, etc.

furthermore, limitations of the first query (offset and limit) cannot be simply adopted in the second query; this is disastrous for the second query if the limit is a small part of the first query results. in this approach, limiting the results of the first query will make it into the second query.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Oct 2007 21:57 UTC**
of course fixing this with LEFT JOIN will solve the pagination issues as well, only that it needs some more analysis and it would only work with HQL

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 Nov 2007 17:31 UTC**
this is not specific to HQL, it's just that presently HQL is the only QL we support that supports subqueries.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 Aug 2008 10:34 UTC**
here's an example of embedded mak:lists: (lists A, B, C, with B and C embedded in A)

<mak:list from="fA" where="wA" >
    <mak:list from="fB" where="wB" >
    ...
    </mak:list>
    <mak:list from="fC" where="wC" >
    ...
    </mak:list>
</mak:list>

it is theoretically possible to do both A and B in the same query, using a LEFT-JOIN. it is also possible to do A and C. however
- doing A, B and C in the same query looks hard to me
- even when doing A+B in the same query, the number of results of A (needed for limiting, pagination) will not be known unless A is executed separately

therefore using LEFT JOIN to correlate A and B or A and C seems NOT to be a good idea
----
using LEFT JOIN for nullable pointers is still a good idea and will be taken care of in bug 1029. 
----

so i change the synopsis of this bug to focus on comment 3. 
correlation of mak:lists A and B is now done like
wA AND wB

but in case A is paginated/limited, the above condition returns all the results of A, not just the ones on the page (the extra A results (and their B counterparts) are filtered out at makumba level). then a better mak:list correlation condition for B is:

wB and fAlabel IN (results of A) 

to be able to do this in an efficient way, it's good to fix bug 1072 first

of course if there are a lot of results from A (like possibly in non-limited A mak:lists) using wA AND wB for mak:list correlation is better. so maybe if there's no LIMIT we should simply use the current mechanism

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 Aug 2008 15:54 UTC**
of course subqueries are also an interesting candidate here

so one would generate (on the example from comment 6)

SELECT makValuesA, (SELECT makValuesB FROM fB WHERE wB), (SELECT makValuesC FROM fC WHERE wC) WHERE wA

i've just done some experiments with such subqueries by inserting in a mak:list
<mak:value expr="(SELECT makValuesB FROM fB WHERE wB)" /> 

the performance is not very different at least in my case (acutally it takes 1 ms extra per subquery, measured at mysql level!)

what i noticed is that when i reduced the number of columns of a query (though not taking away subqueries, but some text fields) the query time halved. so it's clear that the query time depends a lot on the size of the data, so if a lot of extra data is generated by not paginating embedded lists, there is potential for a lot of embedded-mak-list-result garbage. so efforts here are pretty much needed i guess

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 21 Aug 2008 09:28 UTC**
sorry subqueries are not a solution for embedded mak:lists because mysql (and probably others) do not allow a subquery to return more than 1 result unless ANY, ALL, SOME, IN, COUNT, etc are used.

so IN(results of outer query) remains the only solution for this

i suggest to use that only if the outer mak:list is limited. otherwise the outer query may have many-many results so repeating the outer mak:list WHERE condition may be more efficient

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 21 Aug 2008 09:29 UTC**
*** Bug 751 has been marked as a duplicate of this bug. ***

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
